sql: add API for accessing statement diagnostics

dhartunian opened the pull request:
Added 3 new RPC endpoints under `_status` for manipulating Statement Diagnostics
Requests:

1. `CreateStatementDiagnosticsRequest`
Creates a new request for a given statement fingerprint

2. `StatementDiagnosticsRequests`
Fetches all statement diagnostics requests

3. `StatementDiagnostics`
Fetches a specific statement diagnostics request that has been completed and
includes the JSON payload of the trace

Basic implementations for these endpoints have been added to the
`apiReducers.tsx` and `api.ts` files for use by the Admin UI.

A minor bug where completed requests did not have `completedAt` timestamp set
was fixed.

Part of #45246

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45533)
<!-- Reviewable:end -->

dhartunian committed:
sql: add API for accessing statement diagnostics

Added 3 new RPC endpoints under `_status` for manipulating Statement Diagnostics
Requests:

1. `CreateStatementDiagnosticsRequest`
Creates a new request for a given statement fingerprint

2. `StatementDiagnosticsRequests`
Fetches all statement diagnostics requests

3. `StatementDiagnostics`
Fetches a specific statement diagnostics request that has been completed and
includes the JSON payload of the trace

Basic implementations for these endpoints have been added to the
`apiReducers.tsx` and `api.ts` files for use by the Admin UI.

A minor bug where completed requests did not have `completedAt` timestamp set
was fixed.

Release note: none

dhartunian commented:
@koorosh Please use this updated PR as the base to build out the UI. It contains slight modifications since the last draft you saw and adds an implementation of the endpoint to retrieve a completed trace.

RaduBerinde reviewed (approved):
Very cool!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M1TrPnw8MoBIKSmI6sn:b-jdo7t9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/server/serverpb/status.proto, line 781 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1TpV5_9eop6UGn_Kaj:-M1TpV5_9eop6UGn_Kak:b-wo8ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/serverpb/status.proto#L781)):*
> ```protobuf
> }
> 
> message StatementDiagnosticsPayload {
> ```

[nit] this could be just `StatementDiagnostics`
___
*[pkg/server/serverpb/status.proto, line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1Trd5W6hXAIFEU94Oy:-M1Trd5W6hXAIFEU94Oz:b-2zfzn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/serverpb/status.proto#L786)):*
> ```protobuf
>   string trace_json = 3 [(gogoproto.customname) = "TraceJSON"];
>   google.protobuf.Timestamp requested_at = 4
>   [ (gogoproto.nullable) = false, (gogoproto.stdtime) = true ];
> ```

[nit] indent these `[...]` to make it easier to read
___
*[pkg/server/serverpb/status.proto, line 787 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1TqBvuEs4hLDuVI63-:-M1TqBvuEs4hLDuVI630:bmq32y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/serverpb/status.proto#L787)):*
> ```protobuf
>   google.protobuf.Timestamp requested_at = 4
>   [ (gogoproto.nullable) = false, (gogoproto.stdtime) = true ];
>   google.protobuf.Timestamp completed_at = 5
> ```

[nit] I'd use `collected_at`
___
*[pkg/server/serverpb/status.proto, line 789 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1TpZ_5CJ3M6P1NtTe_:-M1TpZ_5CJ3M6P1NtTea:b-smbpws) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/serverpb/status.proto#L789)):*
> ```protobuf
>   google.protobuf.Timestamp completed_at = 5
>   [ (gogoproto.nullable) = true, (gogoproto.stdtime) = true ];
> }
> ```

We should include an `id` here. This`id` will be used to form a URL to download the bundle.

For reference, this is how I did it in the prototype I had: https://github.com/cockroachdb/cockroach/commit/40a56024c2ba29317718a3df72645b730613db77
(full branch here: https://github.com/cockroachdb/cockroach/compare/master...RaduBerinde:bundle-prototype)
I manually set up a `/_admin/v1/stmtbundle` endpoint which serves bundles as files.

Note that the same URL will be given to the user if they use an `EXPLAIN BUNDLE` statement (so it should be reasonably simple).
___
*[pkg/server/serverpb/status.proto, line 976 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1TpGyK2hFIFqjvwcuz:-M1TpGyL6Cn2DGDK0kIy:b-oiw2dj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/serverpb/status.proto#L976)):*
> ```protobuf
>   rpc CreateStatementDiagnosticsRequest(CreateStatementDiagnosticsRequestRequest) returns (CreateStatementDiagnosticsRequestResponse) {
>     option (google.api.http) = {
>       post: "/_status/statementdiagnosticsrequests"
> ```

can we make these shorter? `stmtdiag*` should work just as well
___
*[pkg/sql/statement_diagnostics.go, line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1TqUB9AQNKf8bMGhjT:-M1TqUB9AQNKf8bMGhjU:b3i1ke0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L119)):*
> ```Go
> 			User: security.RootUser,
> 		},
> 		"SELECT sdr.statement_fingerprint, sd.trace, sdr.requested_at, sdr.completed_at "+
> ```

There will be another way to collect a statement diagnostic (via EXPLAIN BUNDLE) and those won't have entries in `statement_diagnostics_requests`. Why can't we only look in `statement_diagnostics`? Just for the requested timestamp? (we could easily add that to `statement_diagnostics`).

Or should I add a completed request whenever one is created via EXPLAIN BUNDLE? CC @andreimatei


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Just nits and naming from me.

There's something about an API which has a key object called a `Request` that you can then later read that leads to minor head explosion when reading. Have you considered calling them something other than a `Request`? How about a `StatementDiagnosticsReport` instead of a `StatementDiagnosticsRequest`. Yes you need to request a report to be generated but at least we won't have to deal with `CreateStatementDiagnosticsRequestResponse` as a type. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M1VkFANCduowhSa_zQh:bsbm64g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @dhartunian)

___
*[pkg/sql/statement_diagnostics.go, line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1ViY2gAjpG9NOHNQtv:-M1ViY2hAe5hWgejA9hs:b-i4t6zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L119)):*
> ```Go
> 			User: security.RootUser,
> 		},
> 		"SELECT sdr.statement_fingerprint, sd.trace, sdr.requested_at, sdr.completed_at "+
> ```

nit: I've been convinced recently to not sweat extra whitespace in SQL queries in code and to use sqlfmt it and shove it in backticks. 

```
`SELECT
    sdr.statement_fingerprint, sd.trace, sdr.requested_at, sdr.completed_at
FROM
    system.statement_diagnostics_requests AS sdr
    LEFT JOIN system.statement_diagnostics AS sd ON
            sdr.statement_diagnostics_id = sd.id
WHERE
    sdr.id = $1 AND sdr.completed = true;`
```
___
*[pkg/sql/statement_diagnostics.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1VjE2U1kkO4WmVf2gA:-M1VjE2V8B2_BdcnFRtj:brk9dd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L134)):*
> ```Go
> 		Completed: true,
> 	}
> 	queryFingerprint, ok := row[0].(*tree.DString)
> ```

complete nit: you could assign these variables in the `if`s.

```
	if queryFingerprint, ok := row[0].(*tree.DString); ok {
		resp.QueryFingerprint = queryFingerprint.String()
	}

```
___
*[pkg/sql/statement_diagnostics.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1VlUVGDE3Eq5Y_A3lU:-M1Vkpia3zx53v3lSWgI:btsy0xp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L184)):*
> ```Go
> 		}
> 
> 		if row[3] != tree.DNull {
> ```

Do you have any interest in pulling the parsing of these rows, or perhaps instead just their fields into a helper that takes maybe nilable datums?

```
func populateStmtDiagnosticsRequestFromDatums(r *StmtDiagnosticsRequest, id, fingerprint, json, completed, sdi, requestedAt, completedAt tree.Datum) error {
 // if the datum is nil, skip it
 // if the daum is the wrong type, return an error
 // ...
}
```


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/server/statement_diagnostics_requests.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X-g0lBL87yOMuTr8K:-M1X-g0lBL87yOMuTr8L:b-imk1cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/statement_diagnostics_requests.go#L19)):*
> ```Go
> )
> 
> func (s *statusServer) CreateStatementDiagnosticsRequest(
> ```

is there precedent for splitting the `statusServer` code outside of `server/status.go`? If there isn't, I woudln't 
___
*[pkg/server/statement_diagnostics_requests.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X5w9cA6oDX3FoXvVv:-M1X5w9cA6oDX3FoXvVw:b-bp1rml) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/statement_diagnostics_requests.go#L39)):*
> ```Go
> }
> 
> func (s *statusServer) StatementDiagnosticsRequests(
> ```

please add some comment to this (the linter is not yelling at you?)
Mention the tables that it populates.
___
*[pkg/server/statement_diagnostics_requests.go, line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X6kFQ3W0_J5mEtzrU:-M1X6kFQ3W0_J5mEtzrV:b-p3u8w7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/statement_diagnostics_requests.go#L65)):*
> ```Go
> 	req *serverpb.StatementDiagnosticsPayloadRequest,
> ) (*serverpb.StatementDiagnosticsPayloadResponse, error) {
> 	ctx = propagateGatewayMetadata(ctx)
> ```

do we need to worry about authorization here?
___
*[pkg/server/statement_diagnostics_requests.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X67-X8mcuU-JncxmE:-M1X67-Y7NzSx4Mg6sMl:b-bp8def) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/server/statement_diagnostics_requests.go#L68)):*
> ```Go
> 	ctx = s.AnnotateCtx(ctx)
> 
> 	diagnostics, err := s.stmtDiagnosticsRequester.GetCompletedRequestWithDiagnostics(ctx, int(req.RequestId))
> ```

line over 100 chars
___
*[pkg/sql/statement_diagnostics.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X6px72dNyDeglHx8I:-M1X6px72dNyDeglHx8J:bsees0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L48)):*
> ```Go
> 
> type StmtDiagnosticsRequest struct {
> 	Id                     int
> ```

ID here and elsewhere
___
*[pkg/sql/statement_diagnostics.go, line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1TqUB9AQNKf8bMGhjT:-M1X73q_5kj4pxDNoD7_:b-2nbyu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L119)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

There will be another way to collect a statement diagnostic (via EXPLAIN BUNDLE) and those won't have entries in `statement_diagnostics_requests`. Why can't we only look in `statement_diagnostics`? Just for the requested timestamp? (we could easily add that to `statement_diagnostics`).

Or should I add a completed request whenever one is created via EXPLAIN BUNDLE? CC @andreimatei
</blockquote></details>

Yeah, I'd try to only look in `statement_diagnostics`. 

I'd also move this code out of the `stmtDiagnosticsRequestRegistry` and into the status server unless there's a particular need for it to be here.
Same with `GetAllRequests()`.
___
*[pkg/sql/statement_diagnostics.go, line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X7TjD9WbaNEzmsxtq:-M1X7TjD9WbaNEzmsxtr:b-bp8def) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L127)):*
> ```Go
> 	}
> 	if row == nil {
> 		return nil, errors.New("requested a statement diagnostics request that does not exist or is not completed")
> ```

line over 100 chars


<!-- Sent from Reviewable.io -->

dhartunian commented:
@RaduBerinde and @andreimatei Agreed regarding the retrieval of statement diagnostics. I've adjusted the protos to mirror the 2 db tables more accurately and the retrieval endpoint now purely accepts a statement diagnostics ID and returns that object. That makes everything much clearer and simpler. Rest of the nits have been addressed as well.

@andreimatei Thank you so much for noting the lack of auth here. We have recently gated statement info behind an admin check and I've added one here as well. 
I'm not sure what you mean by moving code out of the `stmtDiagnosticsRequestRegistry` since the code is methods on a private struct in the module. 
Additionally, there's some precedent for keeping the methods in a separate file: https://github.com/cockroachdb/cockroach/blob/00ed938ab24ec3add2bfe27bea0d0242783f0052/pkg/server/statements.go#L25

@ajwerner Thanks for the look! I like your naming suggestion so I've adopted it. I was implementing mechanically and couldn't think of a good alternative but I like yours here. I've addressed the nits but in the interest of time am not extracting the deserialization function. I love the look of formatted SQL queries in code and am happy to use them here!

andreimatei reviewed (commented):
nit: it's good to reply to comments inline (in Reviewable), otherwise it's hard to match the replies and the discussion threads diverge. For many it's customary to just click the `Done` button.

You know about `crlfmt`, right? Cause the linter is complaining.

> I'm not sure what you mean by moving code out of the stmtDiagnosticsRequestRegistry since the code is methods on a private struct in the module

Can't that code live in the `StatusServer`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M1am0HbCr3gCEMLeLIe:b-4yh1dm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/server/statement_diagnostics_requests.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1an1nT9IRe6nrdY99G:-M1an1nT9IRe6nrdY99H:bb6ht35) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/server/statement_diagnostics_requests.go#L1)):*
> ```Go
> // Copyright 2014 The Cockroach Authors.
> ```

2020
___
*[pkg/server/statement_diagnostics_requests.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1altrV5IyYv-RZcWtb:-M1altrW1Q3SEmaWcdIy:b-dubdt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/server/statement_diagnostics_requests.go#L45)):*
> ```Go
> }
> 
> // StatementDiagnosticsRequests retrieves all of the statement
> ```

which system table?
___
*[pkg/server/serverpb/status.proto, line 973 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1amP8ABfMd_rTQRboH:-M1amP8ABfMd_rTQRboI:b-a24o72) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/server/serverpb/status.proto#L973)):*
> ```protobuf
>     };
>   }
>   rpc CreateStatementDiagnosticsRequest(CreateStatementDiagnosticsReportRequest) returns (CreateStatementDiagnosticsReportResponse) {
> ```

did you want the mismatch between the method name and the argument type name?
___
*[pkg/sql/statement_diagnostics.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1aovThAGUK7f9DvfNH:-M1aovThAGUK7f9DvfNI:b-eh22k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics.go#L172)):*
> ```Go
> 			User: security.RootUser,
> 		},
> 		`SELECT
> ```

I'm a little nervous about not having pagination here. How do other endpoints do it?
I think getting all the incomplete requests is fine, but I don't know about the completed ones.
At least add a TODO?
___
*[pkg/sql/statement_diagnostics.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1aneEtCBiqCw_gLnG0:-M1aneEtCBiqCw_gLnG1:b8uduhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics.go#L463)):*
> ```Go
> 			"UPDATE system.statement_diagnostics_requests "+
> 				"SET completed = true, completed_at = $1, statement_diagnostics_id = $2 WHERE id = $3",
> 			timeutil.Now(), traceID, req.id)
> ```

I think screwed up by leaving the `completed_at` field in :S 
The `collected_at` field in `statement_diagnostics` should be enough; there's always a row in there for a `completed=true` request.
What do you think of a migration (`sqlmigrations/migrations.go`) to drop the column? I can do it too.

If you leave it, share the timestamp with `statement_diagnostics.collected_at`. You can use the SQL `NOW()` for both and it'll be consistent within a transaction.
___
*[pkg/sql/statement_diagnostics_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1apWJo9pCgYutOFM4x:-M1apWJo9pCgYutOFM4y:b-dcrvil) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics_test.go#L131)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	requests, err := tc.Server(0).ExecutorConfig().(ExecutorConfig).stmtInfoRequestRegistry.GetAllRequests(ctx)
> ```

long line
___
*[pkg/sql/statement_diagnostics_test.go, line 150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1apXbR37Fo99Wb56CW:-M1apXbR37Fo99Wb56CX:b-6d26rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics_test.go#L150)):*
> ```Go
> 	}
> 
> 	requests, err := tc.Server(0).ExecutorConfig().(ExecutorConfig).stmtInfoRequestRegistry.GetAllRequests(ctx)
> ```

long line and below too. Configure your editor to yell at you please :)


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @RaduBerinde)

___
*[pkg/sql/statement_diagnostics.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1aneEtCBiqCw_gLnG0:-M1gOlFT9mR9desaxsvp:b-te801g) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics.go#L463)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think screwed up by leaving the `completed_at` field in :S 
The `collected_at` field in `statement_diagnostics` should be enough; there's always a row in there for a `completed=true` request.
What do you think of a migration (`sqlmigrations/migrations.go`) to drop the column? I can do it too.

If you leave it, share the timestamp with `statement_diagnostics.collected_at`. You can use the SQL `NOW()` for both and it'll be consistent within a transaction.
</blockquote></details>

I'm removing the field in https://reviewable.io/reviews/cockroachdb/cockroach/45770


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
>nit: it's good to reply to comments inline (in Reviewable), otherwise it's hard to match the replies and the discussion threads diverge. For many it's customary to just click the Done button.

Thx for the reminder! will do

>You know about crlfmt, right? Cause the linter is complaining.

Did *not* know about crlfmt, now have it configured :)

>Can't that code live in the StatusServer?

They're in different modules (sql vs server) so I don't think I can move them.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M1gYB9P7E8a4PylVSNE:bicj0k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, and @RaduBerinde)

___
*[pkg/server/statement_diagnostics_requests.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1an1nT9IRe6nrdY99G:-M1gKo-n5otQ1u6uTzvW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/server/statement_diagnostics_requests.go#L1)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

2020
</blockquote></details>

Done.
___
*[pkg/server/statement_diagnostics_requests.go, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1altrV5IyYv-RZcWtb:-M1gKzgqDCQ05x4v7vg4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/server/statement_diagnostics_requests.go#L45)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

which system table?
</blockquote></details>

Done.
___
*[pkg/server/serverpb/status.proto, line 973 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1amP8ABfMd_rTQRboH:-M1gLLYz9yJrEyi_vjSF:b-ss7tb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/server/serverpb/status.proto#L973)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

did you want the mismatch between the method name and the argument type name?
</blockquote></details>

Nope. Done.
___
*[pkg/sql/statement_diagnostics.go, line 127 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X7TjD9WbaNEzmsxtq:-M1gLUl31wOXLl6LP-RR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L127)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

line over 100 chars
</blockquote></details>

Done.
___
*[pkg/sql/statement_diagnostics.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1VjE2U1kkO4WmVf2gA:-M1gLW1w2mW8OLn8ONf8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L134)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

complete nit: you could assign these variables in the `if`s.

```
	if queryFingerprint, ok := row[0].(*tree.DString); ok {
		resp.QueryFingerprint = queryFingerprint.String()
	}

```
</blockquote></details>

Done.
___
*[pkg/sql/statement_diagnostics.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1VlUVGDE3Eq5Y_A3lU:-M1gLjLA-LMLupKiQrmH:blxrsk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L184)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Do you have any interest in pulling the parsing of these rows, or perhaps instead just their fields into a helper that takes maybe nilable datums?

```
func populateStmtDiagnosticsRequestFromDatums(r *StmtDiagnosticsRequest, id, fingerprint, json, completed, sdi, requestedAt, completedAt tree.Datum) error {
 // if the datum is nil, skip it
 // if the daum is the wrong type, return an error
 // ...
}
```
</blockquote></details>

Skipping for now.
___
*[pkg/sql/statement_diagnostics.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1aovThAGUK7f9DvfNH:-M1gLXSD7iIsyW-PvzLu:b-iwmm5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics.go#L172)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm a little nervous about not having pagination here. How do other endpoints do it?
I think getting all the incomplete requests is fine, but I don't know about the completed ones.
At least add a TODO?
</blockquote></details>

Added TODO
___
*[pkg/sql/statement_diagnostics.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1aneEtCBiqCw_gLnG0:-M1gM5_FBaOYd1ec6Sms:b-fsm2oz) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics.go#L463)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think screwed up by leaving the `completed_at` field in :S 
The `collected_at` field in `statement_diagnostics` should be enough; there's always a row in there for a `completed=true` request.
What do you think of a migration (`sqlmigrations/migrations.go`) to drop the column? I can do it too.

If you leave it, share the timestamp with `statement_diagnostics.collected_at`. You can use the SQL `NOW()` for both and it'll be consistent within a transaction.
</blockquote></details>

Adding shared timestamp. Just saw your PR to remove the column.
___
*[pkg/sql/statement_diagnostics_test.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1apWJo9pCgYutOFM4x:-M1gMFP06HZKfV1Suo6J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics_test.go#L131)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

long line
</blockquote></details>

Done.
___
*[pkg/sql/statement_diagnostics_test.go, line 150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1apXbR37Fo99Wb56CW:-M1gMFkw9dV43sFKqO8P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics_test.go#L150)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

long line and below too. Configure your editor to yell at you please :)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> They're in different modules (sql vs server) so I don't think I can move them.

But what functionality from the `statement_diagnostics.go` do your new functions need? Aren't they independent?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M1gbzfj7ke6DqjBepjh:bz85h0l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/sql/statement_diagnostics.go, line 463 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1aneEtCBiqCw_gLnG0:-M1gbrAuBv80klm3YrAX:bagpapv) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb3d45b8235fb53f426606916da4db3ec1f4051/pkg/sql/statement_diagnostics.go#L463)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Adding shared timestamp. Just saw your PR to remove the column.
</blockquote></details>

that other PR is in bors now
___
*[pkg/sql/statement_diagnostics.go, line 138 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1gbZiR9q_N7w1Skovb:-M1gbZiR9q_N7w1Skovc:b-qkqyop) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31fa5131a2a08d47fa7b03b19569bdea60a9a22/pkg/sql/statement_diagnostics.go#L138)):*
> ```Go
> 	if row == nil {
> 		return nil, errors.New(
> 			"requested a statement diagnostic that does not exist",
> ```

include the id
___
*[pkg/sql/statement_diagnostics.go, line 165 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1gbcWT04sHkiVMKBTR:-M1gbcWT04sHkiVMKBTS:bchh96t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31fa5131a2a08d47fa7b03b19569bdea60a9a22/pkg/sql/statement_diagnostics.go#L165)):*
> ```Go
> }
> 
> // TODO(davidh): Add pagination to this request
> ```

nit: final period


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/sql/statement_diagnostics.go, line 138 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1gbZiR9q_N7w1Skovb:-M1gq5-HAu7mi4t0i0hR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31fa5131a2a08d47fa7b03b19569bdea60a9a22/pkg/sql/statement_diagnostics.go#L138)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

include the id
</blockquote></details>

Done.
___
*[pkg/sql/statement_diagnostics.go, line 165 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1gbcWT04sHkiVMKBTR:-M1geRIgBgfxrgog16lV:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/d31fa5131a2a08d47fa7b03b19569bdea60a9a22/pkg/sql/statement_diagnostics.go#L165)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: final period
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> But what functionality from the statement_diagnostics.go do your new functions need? Aren't they independent?

Figured it out, thanks! Code has been moved to `statement_diagnostics_requests.go` and the tests were rewritten in `status_test.go`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M1kavAZ6-s8hYEyfqs2:b8peu1n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

dhartunian commented:
@ajwerner @RaduBerinde Andrei is out today and I'd like to get this merged, does that sound ok? I've addressed all the concerns in his last set of reviews.

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M2-9wnf4xXliSDti0ZH:b-qyabs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/server/statement_diagnostics_requests.go, line 63 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M2-9OTl6TLaJ8m__rOy:-M2-9OTmEKSEmZDHUF13:b-mtcfxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b28f34bd95a14a2eaf234102e06f706f95d76d/pkg/server/statement_diagnostics_requests.go#L63)):*
> ```Go
> // request in the `system.statement_diagnostics_requests` table
> // to trace the next query matching the provided fingerprint.
> func (s *statusServer) CreateStatementDiagnosticsReport(
> ```

[nit] we haven't used the word "report" for anything (in the context of stmt diag). How about `RequestStatementDiagnostics`?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> too

I also like Radu's naming nit. 

> Reviewed 2 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-:-M2-XIDfE-hGu4QXElUo:blxpznf)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/server/statement_diagnostics_requests.go, line 63 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M2-9OTl6TLaJ8m__rOy:-M2-WhKm6wsoTNnPEZzt:b-3q6ihq) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b28f34bd95a14a2eaf234102e06f706f95d76d/pkg/server/statement_diagnostics_requests.go#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] we haven't used the word "report" for anything (in the context of stmt diag). How about `RequestStatementDiagnostics`?
</blockquote></details>

SGTM. Just calling these things `StatementDiagnostics` sounds pretty good. Then I guess you'd call the field, `response.Diagnostics`?


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45533)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @andreimatei, @dhartunian, and @RaduBerinde)

___
*[pkg/server/statement_diagnostics_requests.go, line 63 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M2-9OTl6TLaJ8m__rOy:-M2-fK6w5h2-QXeelhM0:b-8i9ex5) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b28f34bd95a14a2eaf234102e06f706f95d76d/pkg/server/statement_diagnostics_requests.go#L63)):*
> [nit] we haven't used the word "report" for anything (in the context of stmt diag). How about RequestStatementDiagnostics?

> SGTM. Just calling these things StatementDiagnostics sounds pretty good. Then I guess you'd call the field, response.Diagnostics?

Thanks, I like these suggestions. I will merge this for now and follow-up with a rename PR once the frontend is integrated since that's ready to go. Then the rename can proceed a bit more smoothly.
___
*[pkg/sql/statement_diagnostics.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45533#-M1X6px72dNyDeglHx8I:-M2-f4mtD_JpQ4WRUlhY:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e150a28f99adce50fc1b05d91260840037ab442/pkg/sql/statement_diagnostics.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ID here and elsewhere
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

dhartunian commented:
Follow-up is tracked here: #45904
